tests/hmm/hmumu_utils.py

```diff
@@ -1914,6 +1915,33 @@ def get_new_btag_weights_shape(jets, evaluator, era, scalars, pt_cut):
         #print(p_jetWt_down[i][mask_pt_bounds[i]],jets.pt[mask_pt_bounds[i]],jets.eta[mask_pt_bounds[i]],jets.btagDeepB[mask_pt_bounds[i]])
     return eventweight_btag , eventweight_btag_up, eventweight_btag_down
 
+#qgl reweighting
+def get_qglWeights(jets, jet_attrs, ret_mu, func, dataset_name):
+    qglw = NUMPY_LIB.ones(jets.numevents())
+    qglw_up = NUMPY_LIB.ones(jets.numevents())
+    psgen = NUMPY_LIB.ones(jets.numevents())
+    if "pythia" in dataset_name:
+        psgen = NUMPY_LIB.zeros(jets.numevents())
+    leading_jet = jets.select_nth(
+                0, ret_mu["selected_events"], None,
+                attributes=jet_attrs)
```

Comment:
`select_nth` always makes a copy of all the data you ask for in `jet_attrs`, which can be redundant, but I have noticed that it's convenient to use if you don't want to deal with jagged arrays... However, if you start to see the code slowing down, you can be more efficient here by copying just the "partonFlavour", "pt', "eta", "qgl" values.

As an option, QGL could also be computed without external C code using something like the following (also directly available on GPU etc):

```python
def compute_qgl_weight(dataset_name, jets):
  if "herwig" in dataset_name:
    qgl = jets.qgl
    fl_less4 = -0.666978*qgl*qgl*qgl + 0.929524*qgl*qgl -0.255505*qgl + 0.981581
    fl_eq21 = -55.7067*qgl*qgl*qgl*qgl*qgl*qgl*qgl + 113.218*qgl*qgl*qgl*qgl*qgl*qgl -21.1421*qgl*qgl*qgl*qgl*qgl -99.927*qgl*qgl*qgl*qgl + 92.8668*qgl*qgl*qgl -34.3663*qgl*qgl + 6.27*qgl + 0.612992
    per_jet_weight = NUMPY_LIB.ones_like(qgl)
    per_jet_weight[NUMPY_LIB.abs(jets.partonFlavour) < 4] = fl_less4
    per_jet_weight[NUMPY_LIB.abs(jets.partonFlavour) == 21] = fl_eq21
  else:
...
  return per_jet_weight
```

Reply:
Thank you very much for the nice suggestion. Now only "partonFlavour", "pt', "eta", "qgl" are included in ```jet_attrs```